astropy/convolution/convolve.py

```diff
@@ -149,6 +155,26 @@ def convolve(array, kernel, boundary='fill', fill_value=0.,
         raise Exception('array and kernel have differing number of '
                         'dimensions.')
 
+    # anything that's masked must be turned into NaNs for the interpolation.
+    # This requires copying the array_internal
+    array_internal_copied = False
+    if np.ma.is_masked(array):
+        if not array_internal_copied:
+            array_internal = array_internal.copy()
+            array_internal_copied = True
+        array_internal[array.mask] = np.nan
+    if mask is not None:
+        if not array_internal_copied:
+            array_internal = array_internal.copy()
+            array_internal_copied = True
+        # mask != 0 yields a bool mask for all ints/floats/bool
+        array_internal[mask != 0] = np.nan
+    if np.ma.is_masked(kernel):
+        if not array_internal_copied:
+            array_internal = array_internal.copy()
+            array_internal_copied = True
+        kernel_internal[kernel.mask] = np.nan
```

Comment:
For the kernel setting masked elements to 0 should work. I'm not sure if the convolution can handle `nan` in the kernel.

Reply:
For anyone looking at this from the future, ``kernel_internal = kernel.filled()``, i.e. ``kernel.filled(fill_value=kernel.fill_value)``, would be better.